File: livingrimoire start here/LivinGrimoire java/src/Auxiliary_Modules/RegexUtil.java
package Auxiliary_Modules;

import java.awt.Point;
import java.util.ArrayList;
import java.util.Hashtable;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

// returns expression of type theRegex from the string str2Check
public class RegexUtil {
    static Hashtable<enumRegexGrimoire, String> regexDictionary = new Hashtable<>();

    static {
        regexDictionary.put(enumRegexGrimoire.email, "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,6}");
        regexDictionary.put(enumRegexGrimoire.timeStamp, "[0-9]{1,2}:[0-9]{1,2}:[0-9]{1,2}");
        regexDictionary.put(enumRegexGrimoire.simpleTimeStamp, "[0-9]{1,2}:[0-9]{1,2}");
        regexDictionary.put(enumRegexGrimoire.secondlessTimeStamp, "[0-9]{1,2}:[0-9]{1,2}");
        regexDictionary.put(enumRegexGrimoire.fullDate, "[0-9]{1,4}/[0-9]{1,2}/[0-9]{1,2} [0-9]{1,2}:[0-9]{1,2}:[0-9]{1,2}");
        regexDictionary.put(enumRegexGrimoire.date, "[0-9]{1,4}/[0-9]{1,2}/[0-9]{1,2}");
        regexDictionary.put(enumRegexGrimoire.double_num, "[-+]?[0-9]*[.,][0-9]*");
        regexDictionary.put(enumRegexGrimoire.integer, "[-+]?[0-9]{1,13}");
        regexDictionary.put(enumRegexGrimoire.repeatedWord, "\\b([\\w\\s']+) \\1\\b");
        regexDictionary.put(enumRegexGrimoire.phone, "[0]\\d{9}");
        regexDictionary.put(enumRegexGrimoire.trackingID, "[A-Z]{2}[0-9]{9}[A-Z]{2}");
        regexDictionary.put(enumRegexGrimoire.IPV4, "([0-9].){4}[0-9]*");
        regexDictionary.put(enumRegexGrimoire.domain, "[A-Za-z0-9.-]+\\.[A-Za-z]{2,6}");
        regexDictionary.put(enumRegexGrimoire.number, "\\d+(\\.\\d+)?");
        regexDictionary.put(enumRegexGrimoire.duplicateWord, "\\b(\\w+)\\b(?=.*\\b\\1\\b)");
        regexDictionary.put(enumRegexGrimoire.firstWord, "^\\w+");
        regexDictionary.put(enumRegexGrimoire.lastWord, "\\w+$");
        regexDictionary.put(enumRegexGrimoire.surname, "\\s+[^\\s]+");
        regexDictionary.put(enumRegexGrimoire.realNumber, "[-+]?[0-9]*[.,][0-9]*"); // -30.77 / 40.05
    }

    public static String extractRegex(String theRegex, String str2Check) {
        Pattern checkRegex = Pattern.compile(theRegex);
        Matcher regexMatcher = checkRegex.matcher(str2Check);
        while (regexMatcher.find()) {
            if (regexMatcher.group().length() != 0) {
                return regexMatcher.group().trim();
            }
        }
        return "";
    }

    public static String extractRegex(enumRegexGrimoire theRegex, String str2Check) {
        Pattern checkRegex = Pattern.compile(regexDictionary.get(theRegex));
        Matcher regexMatcher = checkRegex.matcher(str2Check);
        while (regexMatcher.find()) {
            if (regexMatcher.group().length() != 0) {
                return regexMatcher.group().trim();
            }
        }
        return "";
    }

    public static ArrayList<String> extractAllRegexes(String theRegex, String str2Check) {
        // return a list of all matches
        ArrayList<String> list = new ArrayList<>();
        Pattern checkRegex = Pattern.compile(theRegex);
        Matcher regexMatcher = checkRegex.matcher(str2Check);
        while (regexMatcher.find()) {
            if (regexMatcher.group().length() != 0) {
                list.add(regexMatcher.group().trim());
            }
        }
        return list;
    }

    @SuppressWarnings("unused")
    public static ArrayList<String> extractAllRegexes(enumRegexGrimoire theRegex, String str2Check) {
        // return a list of all matches
        ArrayList<String> list = new ArrayList<>();
        Pattern checkRegex = Pattern.compile(regexDictionary.get(theRegex));
        Matcher regexMatcher = checkRegex.matcher(str2Check);
        while (regexMatcher.find()) {
            if (regexMatcher.group().length() != 0) {
                list.add(regexMatcher.group().trim());
            }
        }
        return list;
    }

    public static Point pointRegex(String str2Check) {
        // "[-+]?[0-9]{1,13}(\\.[0-9]*)?" for double numbers
        String theRegex = "[-+]?[0-9]{1,13}";
        Point result = new Point(0, 0);
        Pattern checkRegex = Pattern.compile(theRegex);
        Matcher regexMatcher = checkRegex.matcher(str2Check);
        while (regexMatcher.find()) {
            if (regexMatcher.group().length() != 0) {
                result.y = Integer.parseInt(regexMatcher.group().trim());
            }
        }
        String phase2 = str2Check.replace(result.y + "", "");
        phase2 = extractRegex(enumRegexGrimoire.number, phase2);
        result.x = Integer.parseInt(phase2);
        return result;
    }
}

File: livingrimoire start here/LivinGrimoire Kotlin/kotlin/auxiliary_modules/RegexUtil.kt
package auxiliary_modules

import java.awt.Point
import java.util.*
import java.util.regex.Pattern

// returns expression of type theRegex from the string str2Check
object RegexUtil {
    private var regexDictionary = Hashtable<enumRegexGrimoire, String>()

    init {
        regexDictionary[enumRegexGrimoire.Email] = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,6}"
        regexDictionary[enumRegexGrimoire.TimeStamp] = "[0-9]{1,2}:[0-9]{1,2}:[0-9]{1,2}"
        regexDictionary[enumRegexGrimoire.SimpleTimeStamp] = "[0-9]{1,2}:[0-9]{1,2}"
        regexDictionary[enumRegexGrimoire.SecondlessTimeStamp] = "[0-9]{1,2}:[0-9]{1,2}"
        regexDictionary[enumRegexGrimoire.FullDate] =
            "[0-9]{1,4}/[0-9]{1,2}/[0-9]{1,2} [0-9]{1,2}:[0-9]{1,2}:[0-9]{1,2}"
        regexDictionary[enumRegexGrimoire.Date] = "[0-9]{1,4}/[0-9]{1,2}/[0-9]{1,2}"
        regexDictionary[enumRegexGrimoire.Double_num] = "[-+]?[0-9]*[.,][0-9]*"
        regexDictionary[enumRegexGrimoire.Integer] = "[-+]?[0-9]{1,13}"
        regexDictionary[enumRegexGrimoire.RepeatedWord] = "\\b([\\w\\s']+) \\1\\b"
        regexDictionary[enumRegexGrimoire.Phone] = "[0]\\d{9}"
        regexDictionary[enumRegexGrimoire.TrackingID] = "[A-Z]{2}[0-9]{9}[A-Z]{2}"
        regexDictionary[enumRegexGrimoire.IPV4] = "([0-9].){4}[0-9]*"
        regexDictionary[enumRegexGrimoire.Domain] = "[A-Za-z0-9.-]+\\.[A-Za-z]{2,6}"
        regexDictionary[enumRegexGrimoire.Number] = "\\d+(\\.\\d+)?"
        regexDictionary[enumRegexGrimoire.DuplicateWord] = "\\b(\\w+)\\b(?=.*\\b\\1\\b)"
        regexDictionary[enumRegexGrimoire.FirstWord] = "^\\w+"
        regexDictionary[enumRegexGrimoire.LastWord] = "\\w+$"
        regexDictionary[enumRegexGrimoire.Surname] = "\\s+[^\\s]+"
        regexDictionary[enumRegexGrimoire.RealNumber] = "[-+]?[0-9]*[.,][0-9]*" // -30.77 / 40.05
    }

    fun extractRegex(theRegex: String, str2Check: String): String {
        val checkRegex = Pattern.compile(theRegex)
        val regexMatcher = checkRegex.matcher(str2Check)
        while (regexMatcher.find()) {
            if (regexMatcher.group().isNotEmpty()) {
                return regexMatcher.group().trim { it <= ' ' }
            }
        }
        return ""
    }

    fun extractRegex(theRegex: enumRegexGrimoire, str2Check: String): String {
        val checkRegex = Pattern.compile(regexDictionary[theRegex]!!)
        val regexMatcher = checkRegex.matcher(str2Check)
        while (regexMatcher.find()) {
            if (regexMatcher.group().isNotEmpty()) {
                return regexMatcher.group().trim { it <= ' ' }
            }
        }
        return ""
    }

    fun extractAllRegexes(theRegex: String, str2Check: String): ArrayList<String> {
        // return a list of all matches
        val list = ArrayList<String>()
        val checkRegex = Pattern.compile(theRegex)
        val regexMatcher = checkRegex.matcher(str2Check)
        while (regexMatcher.find()) {
            if (regexMatcher.group().isNotEmpty()) {
                list.add(regexMatcher.group().trim { it <= ' ' })
            }
        }
        return list
    }

    @Suppress("unused")
    fun extractAllRegexes(theRegex: enumRegexGrimoire, str2Check: String): ArrayList<String> {
        // return a list of all matches
        val list = ArrayList<String>()
        val checkRegex = Pattern.compile(regexDictionary[theRegex]!!)
        val regexMatcher = checkRegex.matcher(str2Check)
        while (regexMatcher.find()) {
            if (regexMatcher.group().isNotEmpty()) {
                list.add(regexMatcher.group().trim { it <= ' ' })
            }
        }
        return list
    }

    fun pointRegex(str2Check: String): Point {
        // "[-+]?[0-9]{1,13}(\\.[0-9]*)?" for double numbers
        val theRegex = "[-+]?[0-9]{1,13}"
        val result = Point(0, 0)
        val checkRegex = Pattern.compile(theRegex)
        val regexMatcher = checkRegex.matcher(str2Check)
        while (regexMatcher.find()) {
            if (regexMatcher.group().isNotEmpty()) {
                result.y = regexMatcher.group().trim { it <= ' ' }.toInt()
            }
        }
        var phase2 = str2Check.replace(result.y.toString() + "", "")
        phase2 = extractRegex(enumRegexGrimoire.Number, phase2)
        result.x = phase2.toInt()
        return result
    }
}